dbt/include/snowflake/macros/materializations/incremental.sql

```diff
@@ -58,8 +58,9 @@
     {% do adapter.expand_target_column_types(
            from_relation=tmp_relation,
            to_relation=target_relation) %}
-    {% do process_schema_changes(on_schema_change, tmp_relation, existing_relation) %}
-    {% set dest_columns = adapter.get_columns_in_relation(existing_relation) %}
+    {% set schema_changes_dict = process_schema_changes(on_schema_change, tmp_relation, existing_relation) %}
+    {#-- Destination columns is the intersection of source and target table --#}
+    {% set dest_columns = schema_changes_dict.get('in_target_and_source', existing_columns) %}
```

Comment:
Only run the extra metadata query if `schema_changes_dict` doesn't return anything (i.e. `on_schema_change: ignore`):
```suggestion
    {% set dest_columns = schema_changes_dict.get('in_target_and_source') %}
    {% if not dest_columns %}
      {% set dest_columns = adapter.get_columns_in_relation(existing_relation) %}
    {% endif %}
```